haskell-lsp-types/src/Language/Haskell/LSP/Types/DataTypesJSON.hs

```diff
@@ -1093,6 +1093,100 @@ type UnregisterCapabilityRequest = RequestMessage ServerMethod UnregistrationPar
 
 type UnregisterCapabilityResponse = ResponseMessage ()
 
+-- ---------------------------------------------------------------------
+
+-- /**
+--  * Describe options to be used when registering for file system change events.
+--  */
+-- export interface DidChangeWatchedFilesRegistrationOptions {
+-- 	/**
+-- 	 * The watchers to register.
+-- 	 */
+-- 	watchers: FileSystemWatcher[];
+-- }
+--
+-- export interface FileSystemWatcher {
+-- 	/**
+-- 	 * The  glob pattern to watch.
+-- 	 *
+-- 	 * Glob patterns can have the following syntax:
+-- 	 * - `*` to match one or more characters in a path segment
+-- 	 * - `?` to match on one character in a path segment
+-- 	 * - `**` to match any number of path segments, including none
+-- 	 * - `{}` to group conditions (e.g. `**​/*.{ts,js}` matches all TypeScript and JavaScript files)
+-- 	 * - `[]` to declare a range of characters to match in a path segment (e.g., `example.[0-9]` to match on `example.0`, `example.1`, …)
+-- 	 * - `[!...]` to negate a range of characters to match in a path segment (e.g., `example.[!0-9]` to match on `example.a`, `example.b`, but not `example.0`)
+-- 	 */
+-- 	globPattern: string;
+--
+-- 	/**
+-- 	 * The kind of events of interest. If omitted it defaults
+-- 	 * to WatchKind.Create | WatchKind.Change | WatchKind.Delete
+-- 	 * which is 7.
+-- 	 */
+-- 	kind?: number;
+-- }
+--
+-- export namespace WatchKind {
+-- 	/**
+-- 	 * Interested in create events.
+-- 	 */
+-- 	export const Create = 1;
+--
+-- 	/**
+-- 	 * Interested in change events
+-- 	 */
+-- 	export const Change = 2;
+--
+-- 	/**
+-- 	 * Interested in delete events
+-- 	 */
+-- 	export const Delete = 4;
+-- }
+
+data DidChangeWatchedFilesRegistrationOptions =
+  DidChangeWatchedFilesRegistrationOptions {
+    _watchers :: List FileSystemWatcher
+  } deriving (Show, Read, Eq)
+
+data FileSystemWatcher =
+  FileSystemWatcher {
+    _globPattern :: String,
+    _kind :: Maybe WatchKind
+  } deriving (Show, Read, Eq)
+
+data WatchKind =
+  WatchKind {
+    -- | Watch for create events
+    _watchCreate :: Bool,
+    -- | Watch for change events
+    _watchChange :: Bool,
+    -- | Watch for delete events
+    _watchDelete :: Bool
+  } deriving (Show, Read, Eq)
+
+instance A.ToJSON WatchKind where
+  toJSON wk = A.Number (createNum + changeNum + deleteNum)
+    where
+      createNum = if _watchCreate wk then 1 else 0
+      changeNum = if _watchChange wk then 2 else 0
+      deleteNum = if _watchDelete wk then 4 else 0
+
+instance A.FromJSON WatchKind where
+  parseJSON (A.Number n)
+    | 1 <= n && n <= 7 =
+      let
+        (wkDelete, n1) = if 4 <= n then (True, n-4) else (False, n)
+        (wkChange, n2) = if 2 <= n1 then (True, n1-2) else (False, n1)
+        (wkCreate, _) = if 1 <= n2 then (True, n2-1) else (False, n2)
+      in
```

Comment:
Really ugly, but I could not think of a way without introducing loads of types and boilerplate.

Reply:
This looks like a bit mask, so a common way would be with a [bitwise and](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bits.html#v:.-38-.)

Something like

```haskell
parseJSON (A.Number n) = pure $ WatchKind (n .&. 1) (n .&. 2) (n .&. 4)
```

Reply:
Agreed, however, `Scientific` doesnt implement `Bits` :/ we could translate it to Int, but I think that is unsafe.

Reply:
Argh I forgot about Javascript numbers. Surely anything thats not an integer is invalid anyway right? And should fail parsing

Reply:
Given it should be an integer, you can probably use one of the conversion functions to integer, then bitwise it.

Reply:
`testBit` is the function to use!